server/src/main.rs

```diff
@@ -0,0 +1,156 @@
+extern crate hyper;
+
+use hyper::{service, Request, Response, Body, Server, StatusCode};
+use hyper::header::{HeaderMap, HeaderName, HeaderValue};
+use futures::{future::{self, Either}, Future, Stream};
+use serde_json::json;
+use lazy_static::lazy_static;
+use ed25519_dalek::{PublicKey, Signature};
+use std::{env, fs};
+use hpos_state_core::state::State;
+use base64::decode_config;
+
+lazy_static! {
+    static ref X_HPOS_ADMIN_SIGNATURE: HeaderName = HeaderName::from_lowercase(b"x-hpos-admin-signature").unwrap();
+    static ref X_ORIGINAL_URI: HeaderName = HeaderName::from_lowercase(b"x-original-uri").unwrap();
```

Comment:
This can only be URL rather than more generic URI, so it would be better to be more specific here:

```suggestion
    static ref X_ORIGINAL_URL: HeaderName = HeaderName::from_lowercase(b"x-original-url").unwrap();
```

Reply:
Well, nginx is capable of operating on more general notion of URIs and `X_ORIGINAL_URI` is a standard header in traffic proxying. I do not see the reason for semantically limiting our functionality. That said I'd much rather stay with `X_ORIGINAL_URI` header here.

Reply:
Well, I've seen `x-original-url` being used more frequently than `x-original-uri`, and Google results seem to back it up: https://www.google.com/search?q=%22x-original-url%22 (92k results) vs https://www.google.com/search?q=%22x-original-uri%22 (5.4k results). That doesn't matter much either way, so nevermind.